core/runtimes/FiberRuntime/index.js

```diff
@@ -116,97 +118,64 @@ FiberRuntime.prototype._createFiber = function (process) {
+  return !_.some(this._processList, function (process) {
+    return process.cbpending || process.status == Process.Status.ACTIVE
+  });
 };
 
 FiberRuntime.prototype._genInitialQueue = function () {
   var self = this;
   var queue = [];
-  //console.log(self._list);  //xxx
-  //console.log(Object.keys(self._list).length);
-  // A process is selfstarting if its incoming ports are only connected to IIPs
- /* ---------------------------------
-  self._list.forEach(function(process) {
-<<<<<<< HEAD
-	console.log(process);  //xxx
-    var selfstarting = process.inports.reduce(function(currentlySelfstarting, inport) {
-      return currentlySelfstarting && (inport[1].conn instanceof IIPConnection);
-    }, true);
-    
-=======
 
+  // A process is selfstarting if its incoming ports are only connected to IIPs
+  _.forEach(self._processList, function (process) {
     var selfstarting = true;
-    process.inports.forEach(function(inPort) {
-      selfstarting = selfstarting && inPort[1].conn instanceof IIPConnection;
+    _.forEach(process.inports, function (inport) {
+      selfstarting = selfstarting && (inport.conn instanceof IIPConnection);
     });
 
->>>>>>> afac49158bea9b9df1715e8ec67485780e5e422d
-    if(selfstarting) {
-        queue.push(process);
-      }
-      */
-   // -----------------------------
-    var selfstarting = true;
-    for (var key in self._list) {
-    	  if (!(self._list.hasOwnProperty(key)))
-    	      continue; 
-    	  //console.log(self._list[key]);
-    	  selfstarting = true;
-    	  self._list[key].inports.forEach(function(port) {
-    		    if (!(port[1].conn instanceof IIPConnection ))
-    		       selfstarting = false;
-    		  } , this);   	 
-    	 
-          if(selfstarting) {
-             queue.push(self._list[key]);
-          }
+    if (selfstarting) {
+      queue.push(process);
     }
-//--------------------------
-  
+  });
 
   return queue;
-} 
+};
 
 
+var ProcState = Enum([
+  "UPSTREAM_CLOSED",
+  "NO_DATA",
+  "DATA_AVAILABLE"
+]);
 FiberRuntime.prototype._procState = function (proc) {
-  // return 2 if all upstream processes closed down, 1 if no data in
-  // connections, 0 otherwise
-  var hasData = false;
   var allDrained = true;
-  for (var i = 0; i < proc.inports.length; i++) {
-    if (proc.inports[i][1].conn instanceof IIPConnection)
-      continue;
-    allDrained = allDrained && proc.inports[i][1].conn.usedslots == 0
-      && proc.inports[i][1].conn.closed;
-    hasData = hasData || proc.inports[i][1].conn.usedslots > 0;
-  }
-  if (allDrained)
-    return 2;
-  if (!hasData)
-    return 1;
-  return 0;
+  var hasData = false;
+
+  _.forEach(proc.inports, function(port) {
+    var connection = port.conn;
+    if (connection instanceof IIPConnection) {
+      return;
+    }
+
+    allDrained = allDrained && connection.usedslots == 0 && connection.closed;
+    hasData = hasData || connection.usedslots > 0;
+  });
+
+  return  allDrained  ? ProcState.UPSTREAM_CLOSED
```

Comment:
I know this may look a little funky, but it's basically a lookup table.

```
  <condition a> ? <return value A>
: <condition b> ? <return value B>
:                 <default value>
```